v4.2: switch custom form control

gijsbotje opened the pull request:
This adds a neat styling feature for the custom form controls.
Markup stays the same as the checkbox but displays it as a toggle switch.

![schermafbeelding 2017-07-06 om 11 25 55](https://user-images.githubusercontent.com/7714133/27904729-f4673888-623d-11e7-9d14-287811b83683.png)

https://codepen.io/gijsbotje/pen/rwKMmy

https://deploy-preview-23004--twbs-bootstrap4.netlify.com/docs/4.1/components/forms/#switches

gijsbotje committed:
Switch custom form control.

skmedix commented on docs/4.0/components/forms.md:
```diff
@@ -1004,6 +1004,23 @@ Custom checkboxes and radios are inline to start. Add a parent with class `.cust
 </div>
 {% endexample %}
 
+### switches
```
First letter should be capitalized

gijsbotje replied:
fixed

gijsbotje commented:
fixes #22090

mdo commented:
Dope! Thank you for jumping in on this. I'll take a look at it for our v4.1 ship :).

gijsbotje commented:
@mdo @XhmikosR @andresgalante I messed up here :( and cant figure out how to revert it, any suggestions?
Sorry for this... don't understand what wen't wrong

XhmikosR commented:
@gijsbotje: you can just do `git rebase -i upstream/v4-dev` and clean up your local branch.

gijsbotje commented:
maybe i did something wrong, but it didn't work...

XhmikosR commented:
@gijsbotje: you had messed up your branch. Do not merge from upstream. Just rebase it, if needed.

Now, I cleaned it up and force pushed it. So, make sure you do `git fetch origin && git reset --hard origin/custom-switch-form-control` before anything else otherwise you will overwrite the cleanup.

gijsbotje commented:
@XhmikosR thanks dude, will do
i tried to make a pr for a different issue and my forked v4-dev branch is messed up i think
i'll see if i can fix it all

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -1013,6 +1013,32 @@ Custom checkboxes and radios can also be disabled. Add the `disabled` boolean at
+{% example html %}
+<div class="custom-control custom-switch custom-switch-animated">
+  <input type="checkbox" class="custom-control-input" id="customSwitchAnimated">
+  <label class="custom-control-label" for="customSwitchAnimated">toggle this switch element</label>
```
Probably want to uppercase the "Toggle" text

gijsbotje replied:
fixed :)

m5o commented:
@gijsbotje / @mdo the solution to modifying the custom-checkbox with `.custom-checkbox-switch` is super handy. 
http://wingman.mediumra.re/components-bootstrap.html#forms

m5o commented on docs/4.0/components/forms.md:
```diff
@@ -1118,6 +1118,32 @@ Custom checkboxes and radios can also be disabled. Add the `disabled` boolean at
+Optionally you can add the class `.custom-switch-animated` to add a transition to the indicator.
+
+{% example html %}
+<div class="custom-control custom-switch custom-switch-animated">
```
`.custom-switch-animated` is not defined.

gijsbotje replied:
have a look at https://github.com/twbs/bootstrap/pull/23004/files#diff-e5540857a54676b43146ad92add2dbdeR182
it's actually is defined as a nested sass selector

wolfy1339 replied:
They don't like that kind of stuff. Don't use nesting to create new classes

ayoubkhan558-zz reviewed (approved):
I think it should be added to bootstrap, so that we don't need any external library for this purpose.

ayoubkhan558-zz reviewed (changes requested):
Its a good option to have

claudioalmeiida commented:
"mdo removed this from Inbox in v4.2 13 days ago" :(

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control

guledali commented:
@mdo  @XhmikosR Any chance we could see the toggle-switch in this release?

MartijnCuppens commented on scss/_custom-forms.scss:
```diff
@@ -161,6 +161,39 @@
+  }
+
+  &-animated .custom-control-label::after {
+    transition: background-position .1s;
```
I think we don't need a animated and non animated version, you may use the transition in the default version. Also, use the `transition` mixin for this.

XhmikosR commented:
Personally I'm not sure it's ready yet.

MartijnCuppens commented on scss/_variables.scss:
```diff
@@ -542,6 +542,10 @@ $custom-checkbox-indicator-indeterminate-border-color: $custom-checkbox-indicato
 $custom-radio-indicator-icon-checked:           str-replace(url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'%3e%3ccircle r='3' fill='#{$custom-control-indicator-checked-color}'/%3e%3c/svg%3e"), "#", "%23") !default;
 
+$custom-switch-indicator-border-radius:         $custom-control-indicator-size / 2 !default;
+$custom-switch-indicator-icon:                  str-replace(url("data:image/svg+xml;charset=utf8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'%3E%3Ccircle r='3' fill='#{$custom-control-indicator-checked-color}'/%3E%3C/svg%3E"), "#", "%23") !default;
```
It's possible to do this without the base64 icon, just use `border-radius` for the element

MartijnCuppens commented on scss/_custom-forms.scss:
```diff
@@ -161,6 +161,39 @@
+  }
+
+  .custom-control-input:checked ~ .custom-control-label::after {
+    background-position: right center;
```
Don't use background-position for this, use the transform property, this is more performant.

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control

XhmikosR commented:
@gijsbotje: will you be able to address @MartijnCuppens's comments above?

gijsbotje commented:
sorry got pre-occupied with work. i'll try and fix it before the weekend

gijsbotje commented:
i'm struggling a little bit...
as the colour scheme changed, the unchecked state of the switch now doesn't have a gray background. which results into the indicator being not visible until the switch is checked.

to illustrate my issue:
![image](https://user-images.githubusercontent.com/7714133/49182630-13872e80-f35b-11e8-9dcf-9640eca999e8.png)

@MartijnCuppens @XhmikosR any suggestions on what colors to use for what?

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control

XhmikosR commented:
Yeah, that was the issue I noticed too when I said I didn't think this was ready along with transitioning costly properties.

I'll try to have a look tomorrow, but @MartijnCuppens will probably beat me to it.

gijsbotje commented:
would this be ok? i used the color of the border for the indicator
![image](https://user-images.githubusercontent.com/7714133/49184740-70391800-f360-11e8-854c-dd4318e574c6.png)

rafalp commented:
Can you include screenshot of other special control (eg. checkbox) bein on/off/disabled?

Also, is there "disabled + on" state?

gijsbotje commented:
@rafalp i've paired the switch with its checkbox equivalent
![image](https://user-images.githubusercontent.com/7714133/49186145-13d7f780-f364-11e8-96b6-065de95ce42e.png)

thanks for reminding me about the `:checked:disabled` state, forgot about that

gijsbotje committed:
- removed animated modifier and made it standard functionality
- changed the indicator to not use the an svg
- changed animation to use transform
- tweaked height and with calculations of the indicator
- added disabled+checked support
- tweaked scss markup so it's the same as customo radio and checkbox

gijsbotje committed:
Merge remote-tracking branch 'origin/custom-switch-form-control' into custom-switch-form-control

gijsbotje commented:
@MartijnCuppens fixed your comments as best as I could think of.
main question to resolve is if the color for the circle when it's unchecked is ok

rafalp commented:
Those are looking very good, thanks 👍

gijsbotje committed:
Merge branch 'v4-dev' into custom-switch-form-control

MartijnCuppens commented on scss/_custom-forms.scss:
```diff
@@ -161,6 +161,48 @@
+      height: $custom-switch-indicator-size;
+      background-color: $custom-control-indicator-border-color;
+      background-size: $custom-control-indicator-size;
+      border-radius: $custom-switch-indicator-border-radius;
```
This is always `50%` I think?

MartijnCuppens replied:
Forget this, you actually also took care of the ability to use switches that aren't rounded:
![image](https://user-images.githubusercontent.com/11559216/49306616-08efa500-f4d3-11e8-8a33-16aac5f3feee.png)

Sweet!

MartijnCuppens commented on scss/_custom-forms.scss:
```diff
@@ -161,6 +161,48 @@
+      width: $custom-switch-indicator-size;
+      height: $custom-switch-indicator-size;
+      background-color: $custom-control-indicator-border-color;
+      background-size: $custom-control-indicator-size;
```
`background-size` is only relevant for background images which we don't have.

MartijnCuppens commented:
@gijsbotje, the color of the circle looks ok.
IMO the switch looks a bit small, maybe we should try to increase the size a bit? 
It would also be nice to have total control over the track width and height and the spacing between the switch and the label text. With 'the track', I mean this element:
![image](https://user-images.githubusercontent.com/11559216/49306402-5fa8af00-f4d2-11e8-89f2-c10a4e9c81f0.png)

Maybe it's possible to tweak this a bit to so that it's possible to set the following variables:
- The track width
- The track height
- Margin between the switch and the label text
- The size of the circle

If we have control over this in `_variables.scss`, it would be easier to control the look of the switch and it would also be possible to make them look like in this PR:
https://github.com/twbs/bootstrap/pull/26691

rafalp commented:
I would be careful with making switch too big - you'll potentially end with component being too big to use "inline", requiring special positioning or, worse, custom form layout altogether.

XhmikosR committed:
Update _custom-forms.scss

MartijnCuppens commented:
Good point, @rafalp. Forget about size increment than.

rafalp commented:
Yeah, I am thinking now how this component should be sized. Right now its sized same way checkbox is, meaning its interchangeable with it. Perhaps it should also support `.form-control-lg` and `.form-control-sm` for extra variety?

Just dropping ideas - it may be too late for this, or material for new PR bumping extra sizing to custom controls?

gijsbotje commented:
> 💡 Support for contextual classes could be interesting, too.

I would do that as a seperate PR as it also affects the other controls

 

> Yeah, I am thinking now how this component should be sized. Right now its sized same way checkbox is, meaning its interchangeable with it. Perhaps it should also support `.form-control-lg` and `.form-control-sm` for extra variety?
> 
> Just dropping ideas - it may be too late for this, or material for new PR bumping extra sizing to custom controls?

Indeed as you said yourself `material for new PR` 😄 
 

> Maybe it's possible to tweak this a bit to so that it's possible to set the following variables:
> 
> - The track width
> - The track height
> - Margin between the switch and the label text
> - The size of the circle
> If we have control over this in _variables.scss, it would be easier to control the look of the switch and it would also be possible to make them look like in this PR:
> #26691 

I was fiddling around with the variable `custom-control-gutter`, but this isn't really the gutter, right? It doesn't define the space between the indicator and the label, but rather defines the space that the label is pushed over. Then the indicator is pulled back the same amount.
So by default the gutter is set to 1.5rem. This gives the checkbox and radio a breathing room of .5rem. Therefore shouldn't gutter be .5rem? And the left properties set to `$custom-control-indicator-size + $custom-control-gutter` which would result in 1.5rem?

If we implement this change, controlling `the track width` and `space between switch and the label` would be much easier. to get it to be accurate in the current setup i would need to do something like:
`$custom-switch-width + ($custom-control-gutter - $custom-control-indicator-size)`
So if you guys agree with the change to `$custom-control-gutter` I will make a separate PR with those changes. I can imagine this would have some impact on people that make use of the sass, so that would be your call.

What i'm really struggling with and get wrap my head around is making the height and circle size customizable. I'm currently using the calc css function to get the height and with like this:
`calc(#{$custom-control-indicator-size} - #{$custom-control-indicator-border-width * 4})`
First the 1rem indicator-size then subtracting the border width * 4 to create space between the circle and the track. This can not be done in sass math as `$custom-control-indicator-border-width` is a pixel value and `$custom-control-indicator-size` is a rem value, which are incompatible in sass.

i'm also struggling with the positioning of the circle when you increase or decrease the size of the circle. currently this calc function is taking care of the vertical positioning:
`calc(#{(($font-size-base * $line-height-base - $custom-control-indicator-size) / 2)} + #{$custom-control-indicator-border-width * 2})`

I'm far from a math guy so if so if someone could assist me in these issues that would be much appreciated.

gijsbotje committed:
- improved width sizing
- improved gutter calculation

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control

gijsbotje committed:
Merge branch 'v4-dev' into custom-switch-form-control

gijsbotje committed:
Merge branch 'v4-dev' into custom-switch-form-control

MartijnCuppens reviewed (approved):
LGTM, more control over sizing is something we can look into later

XhmikosR committed:
Merge branch 'v4-dev' into custom-switch-form-control
